example/contrib/socks-client/socks/impl/async_handshake_v5.hpp

```diff
@@ -397,6 +404,35 @@ class socks5_op : public base_type, public net::coroutine
     int step_ = 0;
 };
 
+struct async_handshake_v5_initiator
+{
+  template<class HandlerType, class AsyncStream>
+  void operator()(
+    HandlerType&& handler,
+    AsyncStream& stream,
```

Comment:
I'm not certain, but I believe references need special treatment. You might ask in #general . I think Chris passes them as pointers to avoid some ambiguity (although, this could be because he tries to be C++03 compatible).

Reply:
I had a look in the asio code. Chris performs a BOOST_ASIO_MOVE_ARG on each argument. In c++11 or better, this will of course ccast all arguments to r-values which is fine because the composed operation will cast them back to l-value references.
In the obsolete pre-c++11 world there might be some boost-move-related issue, but this is no concern of ours.